domain/learningandworkprogress/src/test/kotlin/uk/gov/justice/digital/hmpps/domain/learningandworkprogress/review/service/ReviewServiceTest.kt

```diff
@@ -79,4 +84,273 @@ class ReviewServiceTest {
       verify(reviewPersistenceAdapter).getCompletedReviews(PRISON_NUMBER)
     }
   }
+
+  @Nested
```

Comment:
There's A LOT of tests here, but it feels like the calculation of the review dates etc for all the different permutations/scenarios needs a lot of testing. And hopefully these tests will help protect us against any regressions as we iterate this code 👍 🤞

Reply:
I feel like this is a candidate for parameterised tests... 

`  @ParameterizedTest
    @MethodSource("testCases")
    fun `it should determine calculation rule given prisoner scenario`(
        releaseDate: LocalDate,
        sentenceType: SentenceType,
        isReAdmission: Boolean,
        isTransfer: Boolean,
        expectedRule: ReviewScheduleCalculationRule
    ) {
        // When
        val actual = service.determineReviewScheduleCalculationRule(releaseDate, sentenceType, isReAdmission, isTransfer)

        // Then
        assertEquals(expectedRule, actual)
    }

    companion object {
        @JvmStatic
        fun testCases(): Stream<Arguments> {
            return Stream.of(
                Arguments.of(
                    LocalDate.now().plusYears(2),
                    SentenceType.SENTENCED,
                    true,
                    false,
                    ReviewScheduleCalculationRule.PRISONER_READMISSION
                ),
                Arguments.of(
                    LocalDate.now().plusYears(2),
                    SentenceType.SENTENCED,
                    false,
                    true,
                    ReviewScheduleCalculationRule.PRISONER_TRANSFER
                ),
                // etc...
            )
        }
    }`

Reply:
Yes, agreed, a parameterised test is much neater 👍 
To make it even neater I've split it out into a test class of it's own because even with the method source, it is stlll a lot of code.
This is how it runs now:
![Screenshot 2024-11-06 at 10 46 43](https://github.com/user-attachments/assets/b614107c-3fa2-46f9-a8ee-391dfe049102)